editor/components/font-sizes/with-font-sizes.js

```diff
@@ -0,0 +1,125 @@
+ *
+ * @return {Function} Higher-order component.
+ */
+export default ( ...fontSizeNames ) => {
+	// computes an object whose key is the font size attribute name as passed in the array,
+	// and the value is the custom font size attribute name.
+	// Custom font size is automatically compted by appending custom followed by the font size attribute name in with the first letter capitalized.
+	const fontSizeAttributeNames = reduce( fontSizeNames, ( fontSizeAttributeNamesAccumulator, fontSizeAttributeName ) => {
+		fontSizeAttributeNamesAccumulator[ fontSizeAttributeName ] = `custom${ upperFirst( fontSizeAttributeName ) }`;
+		return fontSizeAttributeNamesAccumulator;
+	}, {} );
+	return createHigherOrderComponent(
+		compose( [
+			withSelect( ( select ) => {
+				const { fontSizes } = select( 'core/editor' ).getEditorSettings();
+				return {
+					fontSizes,
+				};
+			} ),
+			( WrappedComponent ) => {
+				return class extends Component {
+					constructor( props ) {
+						super( props );
+
+						this.setters = this.createSetters();
+
+						this.state = {};
+					}
+
+					createSetters() {
+						return reduce( fontSizeAttributeNames, ( settersAccumulator, customFontSizeAttributeName, fontSizeAttributeName ) => {
+							const upperFirstFontSizeAttributeName = upperFirst( fontSizeAttributeName );
+							settersAccumulator[ `set${ upperFirstFontSizeAttributeName }` ] =
+								this.createSetFontSize( fontSizeAttributeName, customFontSizeAttributeName );
+							return settersAccumulator;
+						}, {} );
+					}
+
+					createSetFontSize( fontSizeAttributeName, customFontSizeAttributeName ) {
+						return ( fontSizeValue ) => {
+							const fontSizeObject = find( this.props.fontSizes, { size: fontSizeValue } );
+							this.props.setAttributes( {
+								[ fontSizeAttributeName ]: fontSizeObject && fontSizeObject.slug ? fontSizeObject.slug : undefined,
+								[ customFontSizeAttributeName ]: fontSizeObject && fontSizeObject.slug ? undefined : fontSizeValue,
+							} );
+						};
+					}
+
+					static getDerivedStateFromProps( { attributes, fontSizes }, previousState ) {
+						const didAttributesChange = ( customFontSizeAttributeName, fontSizeAttributeName ) => {
+							if ( previousState[ fontSizeAttributeName ] ) {
+								// if new font size is name compare with the previous slug
+								if ( attributes[ fontSizeAttributeName ] ) {
+									return attributes[ fontSizeAttributeName ] !== previousState[ fontSizeAttributeName ].slug;
+								}
+								// if font size is not named, update when the font size value changes.
+								return previousState[ fontSizeAttributeName ].size !== attributes[ customFontSizeAttributeName ];
+							}
+							// in this case we need to build the font size object
+							return true;
+						};
+
+						if ( ! some( fontSizeAttributeNames, didAttributesChange ) ) {
+							return previousState;
+						}
+
```

Comment:
It seems like you could further simplify this code using `filter` method;

```
const changedAttributes = filter( fontSizeAttributeNames, didAttributesChange );

if ( changedAttributes.length === 0 ) {
    return null;
}

const newState  = reduce( fontSizeAttributeNames, ( newStateAccumulator, customFontSizeAttributeName, fontSizeAttributeName ) => {
    // the same logic but without if statement
}, {} );

return {
    ...previousState,
    ...newState,
};
```

Reply:
I did some code improvements following the logic you proposed. I was not able to use filter because filters always return an array but I used pickBy (is equivalent for the object case).
I kept the "some" verification, because creating a new object is still a non-trivial operation and doing the check allows us to on the most common case (font attributes did not change) do only comparisons.